# Regression Example Using TensorFlow 2
## from Geron, 2nd Edition

In [7]:
import tensorflow as tf
from tensorflow import keras

tf.__version__

'2.3.1'

In [8]:
keras.__version__

'2.4.0'

In [9]:
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

housing = fetch_california_housing()

X_train_full, X_test, y_train_full, y_test = train_test_split(
    housing.data, housing.target)
X_train, X_valid, y_train, y_valid = train_test_split(
    X_train_full, y_train_full)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_valid = scaler.transform(X_valid)
X_test = scaler.transform(X_test)

print(housing.data)
print(housing.target)
print(len(housing.data[0]))
print(len(X_train))
print(len(y_train))
X_train.shape[1:]

[[   8.3252       41.            6.98412698 ...    2.55555556
    37.88       -122.23      ]
 [   8.3014       21.            6.23813708 ...    2.10984183
    37.86       -122.22      ]
 [   7.2574       52.            8.28813559 ...    2.80225989
    37.85       -122.24      ]
 ...
 [   1.7          17.            5.20554273 ...    2.3256351
    39.43       -121.22      ]
 [   1.8672       18.            5.32951289 ...    2.12320917
    39.43       -121.32      ]
 [   2.3886       16.            5.25471698 ...    2.61698113
    39.37       -121.24      ]]
[4.526 3.585 3.521 ... 0.923 0.847 0.894]
8
11610
11610


(8,)

### We have 11,610 records of X and y to train from.  The X contains 8 attributes, and the y are single values.

In [10]:
model = keras.models.Sequential([
    keras.layers.Dense(8, activation="relu", input_shape=X_train.shape[1:]),
    keras.layers.Dense(1)
])

model.compile(loss="mean_squared_error", optimizer="sgd")

In [21]:
history = model.fit(X_train, y_train, epochs=20,
                    validation_data=(X_valid, y_valid))

Epoch 1/20
363/363 [==============================] - 0s 990us/step - loss: 0.9125 - val_loss: 0.6547
Epoch 2/20
363/363 [==============================] - 0s 806us/step - loss: 0.5584 - val_loss: 0.5509
Epoch 3/20
363/363 [==============================] - 0s 798us/step - loss: 0.5141 - val_loss: 0.5162
Epoch 4/20
363/363 [==============================] - 0s 753us/step - loss: 0.5164 - val_loss: 0.5121
Epoch 5/20
363/363 [==============================] - 0s 823us/step - loss: 0.4854 - val_loss: 0.4872
Epoch 6/20
363/363 [==============================] - 0s 784us/step - loss: 0.4834 - val_loss: 0.4795
Epoch 7/20
363/363 [==============================] - 0s 795us/step - loss: 0.4829 - val_loss: 0.4762
Epoch 8/20
363/363 [==============================] - 0s 786us/step - loss: 0.4670 - val_loss: 0.4709
Epoch 9/20
363/363 [==============================] - 0s 812us/step - loss: 0.4639 - val_loss: 0.4738
Epoch 10/20
363/363 [==============================] - 0s 811us/step - loss: 0.453

In [22]:
mse_test = model.evaluate(X_test, y_test)
mse_test

162/162 [==============================] - 0s 521us/step - loss: 0.3993


0.39934536814689636

In [23]:
X_new1 = X_test[:3] # pretend these are new instances
y_pred1 = model.predict(X_new1)
y_pred1

array([[5.1036253],
       [2.9212265],
       [2.1466863]], dtype=float32)

In [24]:
X_new2 = X_test[:3] # pretend these are new instances
y_pred2 = model.predict(X_new2)
y_pred2

array([[5.1036253],
       [2.9212265],
       [2.1466863]], dtype=float32)

### Now set up TensorBoard

In [5]:
import os
root_logdir = os.path.join(os.curdir, "my_logs")

def get_run_logdir():
    import time
    run_id = time.strftime("run_%Y_%m_%d-%H_%M_%S")
    return os.path.join(root_logdir, run_id)

run_logdir = get_run_logdir() # e.g., './my_logs/run_2019_06_07-15_15_22'

In [11]:
tensorboard_cb = keras.callbacks.TensorBoard(run_logdir)
history = model.fit(X_train, y_train, epochs=20,
                    validation_data=(X_valid, y_valid),
                    callbacks=[tensorboard_cb])

Epoch 1/20
  1/363 [..............................] - ETA: 0s - loss: 6.9559WARNING:tensorflow:From /usr/local/lib/python3.8/site-packages/tensorflow/python/ops/summary_ops_v2.py:1277: stop (from tensorflow.python.eager.profiler) is deprecated and will be removed after 2020-07-01.
Instructions for updating:
use `tf.profiler.experimental.stop` instead.
363/363 [==============================] - 0s 1ms/step - loss: 0.8852 - val_loss: 0.6128
Epoch 2/20
363/363 [==============================] - 0s 896us/step - loss: 0.4604 - val_loss: 0.8439
Epoch 3/20
363/363 [==============================] - 0s 845us/step - loss: 0.4317 - val_loss: 0.8417
Epoch 4/20
363/363 [==============================] - 0s 851us/step - loss: 0.4218 - val_loss: 0.6234
Epoch 5/20
363/363 [==============================] - 0s 858us/step - loss: 0.4156 - val_loss: 0.5716
Epoch 6/20
363/363 [==============================] - 0s 828us/step - loss: 0.4115 - val_loss: 0.4850
Epoch 7/20
363/363 [===========================